src/debug/debug_stream/debug_stream_slot.h

```diff
@@ -0,0 +1,137 @@
+ *      |   {                                            |   |
+ *      |      core_id = 1                               |   |
+ *      |      size = 1344                               | 64 bytes
+ *      |      offset = 1344+64                          |   |
+ *      |   }                                            |   |
+ *      | <padding>                                      |   |
+ *      --------------------------------------------------  ---
+ *      |   {                                            |   |
+ *      |      core_id = 2                               |   |
+ *      |      size = 1344                               | 64 bytes
+ *      |      offset = 2*1344+64                        |   |
+ *      |   }                                            |   |
+ *      | }                                              |   |
+ *      | <padding>                                      |   |
+ *      --------------------------------------------------  ---
+ *   * CONFIG_MP_MAX_NUM_CPUS is 3 in this example
+ *
+ * The header contains generic information like identifier, total
+ * size, and number of sections. After the generic fields there is an
+ * array of section descriptors. Each array element is cacheline
+ * aligned. The array has 'num_sections' number of elements. Each
+ * element in the array describes a circular buffer, one for each DSP
+ * core.
+ *
+ * The remaining memory in the debug window slot is divided between
+ * those sections. The buffers are not necessarily of equal size, like
+ * in this example. The circular buffers are all cache line aligned,
+ * 64 in this example. One section looks like this:
+ *
+ *      --------------------------------------------------  ---
+ *      | next_serial = <counter for written objects>    |   |
+ *      | w_ptr = <write position in 32-bit words>       | 1344 bytes
+ *      | buffer_data[1344/4-2] = {                      |   |
+ *      |    <debug data records>                        |   |
+ *      | }                                              |   |
+ *      --------------------------------------------------  ---
+ *
+ * The data records are described in debug_strem.h. In the case of
+ * debug window memory slot the next record should always be aligned
+ * to word (4-byte) boundary.
+ *
+ * The debug stream writes the records of abstract data to the
+ * circular buffer, and updates the w_ptr before it writes record.
+ * The host side receiver tries to keep up with the w_ptr and keeps
+ * track of its read position. The size of the record is written -
+ * again - after each record and before the next. This is to allow
+ * parsing the stream backwards in an overrun recovery situation. The
+ * w_ptr value is updated last, when the record is completely written.
+ */
+
+#include <stdint.h>
+
+/* Core specific section descriptor
+ *
+ * Section descriptor defines core ID, offset and size of the circular
+ * buffer in the debug window slot.
+ */
+struct debug_stream_section_descriptor {
```

Comment:
again - is user API, needs to go in another directory and needs packed/aligned.

Reply:
Packed hardly does anything when everything is 32-bit words, and the descriptor is already aligned to cache-line size. The header itself must be written to beginning of slot so its automatically 4k page size aligned, which should be enough, and also if its not the idea is still to have it at the  very beginning of the slot. I can put the packed aligned stuff to the header too, its not needed, and I feel it could even be misleading.

Reply:
best to be explicit here even though parent is packed and this is all 32bit words, i.e. we could have direct users that may reference this structure without the parent context (and hence optimize for 32 but aligned/packed access).

Reply:
@jsarha FWIW agree with @lgirdwood - please add `packed`. It's good for all ABI types to have it even if they "obviously" effectively would be packed by default by the compiler anyway.

Reply:
It looks like if in xcc compilation (I have not tried with gcc) I put __packed to a struct, that contains an array of structs that are __aligned(64), then the first array elemen is not 64-byte aligined, but packed next to previous element, second is then 64-bytes after the first, but still wrongly aligned. 

So I can not make struct debug_stream_slot_hdr __packed, because then the elements of section_desc[] are not 64-byte aligned. And adding __aligned(64) to   too does not help. So at least I have to drop the __packed from struct debug_stream_slot_hdr.

Reply:
> if put __packed to a struct, that contains an array of structs that are __aligned(64), then the first array elemen is not 64-byte aligined, but packed next to previous element, second is then 64-bytes after the first, but still wrongly aligned.

Can provide the C code and `pahole` output? I'm not 100% sure I follow. How can the first element have a "previous" element?

I suspect `aligned(64)` could be automatically downgraded to `aligned(32)` in some 32 bits environments. Quoting https://gcc.gnu.org/onlinedocs/gcc/Common-Type-Attributes.html#index-packed-type-attribute

> On many systems, the linker is only able to arrange for variables to be aligned up to a certain maximum alignment. [...] If your linker is only able to align variables up to a maximum of 8-byte alignment, then specifying aligned (16) in an `__attribute__` still only provides you with 8-byte alignment.

Reply:
@jsarha alignment doesn't always play well with packing. You pack data explicitly to make it "predictable," "reproducible." And it *can well* break alignment. Or at least you take care of alignment yourself. E.g. if you have
```
struct x {
    uint8_t c;
    uint32_t i;
} __attribute__((packed));
```
then sure enough `i` will be misaligned, and if you attempt to access it directly on some architectures you get an alignment violation exception. So you either add padding there or rearrange members manually to achieve desired alignment. So, I'd say that ABI types should be packed and alignment should be provided manually, or you end up using `memcpy()` to access unaligned members.

Reply:
> And it can well break alignment

I think you mean: break "normal" alignment.

I'm relatively confident `packed` is in most cases equivalent to `packed,aligned(1)` where `packed,aligned(X)` means: align on exactly X while `aligned(X)` (not packed) means: aligned on _at least_ X but let the optimizer align higher than X if it thinks it knows better.


> I suspect aligned(64) could be automatically downgraded to aligned(32) in some 32 bits environments.

A clearer statement here: https://www.gnu.org/software/c-intro-and-ref/manual/html_node/Attributes.html
> However, you can’t force an alignment bigger than the computer’s maximum meaningful alignment.

Reply:
I can of course create all kinds of constructs to invent the __aligned() again with variable sized paddings, and use __packed, but now I do not have time for it so, I leave this as it is.

Reply:
@jsarha Wouldn't just removing `__aligned` and using `__packed` already solve it? And just look through those types to verify that you arrange elements from larger to smaller - first all 64-bit ones, then 32-bit ones etc. Only if you embed structures, well, then yes, just pad the structure that you're embedding with some `uint8_t reserved[N]` array to align to the next 64 bits, that's it

Reply:
Yes, there is several ways to do this, and probably it would work even if the core specific  circular-buffer descriptors were not 64-byte aligned, since after the initialization the contents is read only and the memory is accessed through uncached address-space. Still @lgirdwood suggested that also the static descriptor part should be 64-bit aligned, just to be safe. I am really fine with anything that works, but I would like to do it only once more, and the time is running little short if we want this to go to 2.11.

Reply:
@lgirdwood @lyakh , I am more and more starting to thing the __aligned() should be dropped from "struct debug_stream_section_descriptor", and for that matter all the padding hacks too. We can not have the ABI depend on cache-line size, or at least we should then write the offsets explicitly out in the debug window, but that is pretty much what the "struct debug_stream_section_descriptor" is for:  It is all static data that is anyway initialized by core 0 before the threads running on other cores have started and its in uncached memory, and it contains the offsets of the core specific circular buffers, and their sizes. So, there should not and must not be any cash-line size dependent padding in the header that describes where the core-specific circular buffers are.  I've tested this works fine with sof-mtl-nocodec-ssp0-ssp2.tplg on MTL.

Reply:
@jsarha I agree. The padding that I was proposing wasn't for cache-line size alignment, it was for "natural data type alignment." I.e. if you have an 8-bit member followed by a 16-bit value and you cannot swap them, add a padding like
```
struct abi_data {
    uint8_t type;
    uint8_t reserved0[3];
    uint32_t size;
    ....
};
```